keras_nlp/metrics/bleu.py

```diff
@@ -0,0 +1,449 @@
+            return inputs
+
+        if tokenizer is None:
+            self.tokenizer = default_tokenizer
+        else:
+            self.tokenizer = tokenizer
+        self.max_order = max_order
+        self.smooth = smooth
+        self.variant = variant
+
+        if variant == "corpus_bleu":
+            self._matches = self.add_weight(
+                shape=(self.max_order,),
+                name="bleu_matches",
+                initializer="zeros",
+                dtype=self.dtype,
+            )
+            self._possible_matches = self.add_weight(
+                shape=(self.max_order,),
+                name="bleu_possible_matches",
+                initializer="zeros",
+                dtype=self.dtype,
+            )
+            self._translation_length = self.add_weight(
+                name="bleu_translation_length",
+                initializer="zeros",
+                dtype=self.dtype,
+            )
+            self._reference_length = self.add_weight(
+                name="bleu_reference_length",
+                initializer="zeros",
+                dtype=self.dtype,
+            )
+        else:
+            self._number_of_samples = self.add_weight(
+                name="number_of_samples",
+                initializer="zeros",
+                dtype=self.dtype,
+            )
+
+        self._bleu = self.add_weight(
+            name="bleu",
+            initializer="zeros",
+            dtype=self.dtype,
+        )
+
+    def update_state(self, y_true, y_pred, sample_weight=None):
+        def validate_and_fix_rank(inputs, tensor_name, base_rank=0):
+            if not isinstance(inputs, tf.Tensor):
+                inputs = tf.convert_to_tensor(inputs)
+
+            if inputs.shape.rank == base_rank:
+                return inputs[tf.newaxis]
+            elif inputs.shape.rank == base_rank + 1:
+                return inputs
+            else:
+                raise ValueError(
+                    f"{tensor_name} must be of rank {base_rank} or {base_rank+1}. "
+                    f"Found rank: {inputs.shape.rank}"
+                )
+
+        def _get_ngrams(segment, max_order):
+            """Extracts all n-grams upto a given maximum order from an input
+            segment. Uses Python ops. Inspired from
+            https://github.com/tensorflow/nmt/blob/master/nmt/scripts/bleu.py.
+
+            Args:
+                segment: string. Text segment from which n-grams will be
+                    extracted.
+                max_order: int. Maximum length in tokens of the n-grams returned
+                    by this methods.
+            """
+            ngram_counts = collections.Counter()
+            for order in range(1, max_order + 1):
+                for i in range(0, len(segment) - order + 1):
+                    ngram = tuple(segment[i : i + order])
+                    ngram_counts[ngram] += 1
+            return ngram_counts
+
+        def corpus_bleu(
```

Comment:
Why is this public? do we expect this to be called explicitly? If so can you show the use case?